controllers/controller_test.go

```diff
@@ -155,3 +158,87 @@ func TestCorrectsMisconfiguredWf(t *testing.T) {
 		t.Errorf("expected correctedWf=cronWf, but got diff: %s", diff)
 	}
 }
+
+func TestTaskPhaseMatchesWorkflowPhase(t *testing.T) {
+	testTask, err := createTestTask(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wf, err := workflows.WaitForTaskWorkflow(ctx, workflowClient, testTask.GetName(), time.Second)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testTask, err = taskGenClient.Cli.Get(ctx, testTask.GetName(), v1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tcs := []struct {
+		taskPhase     v1alpha1.TaskPhase
+		workflowPhase argo.WorkflowPhase
+	}{
+		{
+			taskPhase:     v1alpha1.TaskRunning,
+			workflowPhase: argo.WorkflowRunning,
+		},
+		{
+			taskPhase:     v1alpha1.TaskPending,
+			workflowPhase: argo.WorkflowPending,
+		},
+		{
+			taskPhase:     v1alpha1.TaskFailed,
+			workflowPhase: argo.WorkflowFailed,
+		},
+		{
+			taskPhase:     v1alpha1.TaskFailed,
+			workflowPhase: argo.WorkflowError,
+		},
+		{
+			taskPhase:     v1alpha1.TaskSucceeded,
+			workflowPhase: argo.WorkflowSucceeded,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(fmt.Sprintf("taskPhase: %s, workflowPhase: %s", tc.taskPhase, tc.workflowPhase), func(t *testing.T) {
+			wf.Status.Phase = tc.workflowPhase
+			wf, err = workflowGenClient.Cli.Update(ctx, wf, v1.UpdateOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			err = taskGenClient.WaitForCondWithTimout(ctx, testTask.GetName(), common.GenTaskPhaseValidator(tc.taskPhase), time.Second)
+			if err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestHandleGitSourceCorrectly(t *testing.T) {
+	source := v1alpha1.NewGitSrc("myrepo", "myref")
+	task := v1alpha1.NewTaskWithSrc("mycmd", "myproject", source)
+	task, err := tasks.Create(ctx, task, v1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wf, err := workflows.WaitForTaskWorkflow(ctx, workflowClient, task.GetName(), time.Second)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wfSetupScript := wf.Spec.Templates[0].Steps[0].Steps[0].Inline.Script.Source
```

Comment:
I think we also extract the Inline wf step in a different test, it may be a good idea to separate this logic into a function.